mne/preprocessing/xdawn.py

```diff
@@ -0,0 +1,565 @@
+                # init sklearn.covariance.OAS estimator
+                skl_cov = OAS(store_precision=False,
+                              assume_centered=True)
+            else:
+                raise ValueError("regularization parameter should be "
+                                 "of type str (got %s)." % type(reg))
+        else:
+            raise ValueError("regularization parameter should be "
+                             "of type str (got %s)." % type(reg))
+
+        # compute regularized covariance using sklearn
+        covar = skl_cov.fit(data.T).covariance_
+
+    return covar
+
+
+def _check_overlapp(epochs):
+    """check if events are overlapped."""
+    isi = np.diff(epochs.events[:, 0])
+    window = int((epochs.tmax - epochs.tmin) * epochs.info['sfreq'])
+    return isi.min() < window
+
+
+def _construct_signal_from_epochs(epochs):
+    """Reconstruct pseudo continuous signal from epochs."""
+    start_ix = (np.min(epochs.events[:, 0])
+                + int(epochs.tmin * epochs.info['sfreq']))
+    end_ix = (np.max(epochs.events[:, 0])
+              + int(epochs.tmax * epochs.info['sfreq']) + 1)
+
+    ns = end_ix - start_ix
+    ns_epochs = epochs._data.shape[2]
+    ne = epochs._data.shape[1]
+    ix_events = epochs.events[:, 0] - epochs.events[0, 0]
+
+    data = np.zeros((ne, ns))
+    for i, ix in enumerate(ix_events):
+        data[:, ix:(ix + ns_epochs)] = epochs._data[i]
+
+    return data
+
+
+def least_square_evoked(epochs, return_toeplitz=False):
+    """Least square estimation of evoked response from a Epochs instance.
+
+    Parameters
+    ----------
+    epochs : Epochs instance
+        An instance of Epochs.
+    return_toeplitz : bool
+        If true, compute the toeplitz matrix.
+
+    Returns
+    -------
+    evokeds : dict of evoked instance
+        An dict of evoked instance for each event type in epochs.event_id.
+    toeplitz : dict of array
+        If return_toeplitz is true, return the toeplitz matrix for each event
+        type in epochs.event_id.
+    """
+    if not isinstance(epochs, _BaseEpochs):
+        raise ValueError('epochs must be an instance of `mne.Epochs`')
+
+    evs = epochs.events.copy()
+    evs[:, 0] -= evs[0, 0] + int(epochs.tmin * epochs.info['sfreq'])
+    data = _construct_signal_from_epochs(epochs)
+    evo, to = _least_square_evoked(data, evs, epochs.event_id,
+                                   tmin=epochs.tmin, tmax=epochs.tmax,
+                                   sfreq=epochs.info['sfreq'], decim=1)
+    evokeds = dict()
+    info = cp.deepcopy(epochs.info)
+    for name, data in evo.items():
+        n_events = len(evs[evs[:, 2] == epochs.event_id[name]])
+        evoked = EvokedArray(data, info, tmin=epochs.tmin,
+                             comment=name, nave=n_events)
+        evokeds[name] = evoked
+
+    if return_toeplitz:
+        return evokeds, to
+
+    return evokeds
+
+
+class Xdawn(TransformerMixin, ContainsMixin):
```

Comment:
To avoid me some pain, please add `versionadded` directive and also `See also` :)